mlir/lib/Dialect/Rock/Transforms/RockMultibuffer.cpp

```diff
@@ -276,25 +296,67 @@ mlir::rock::multiBuffer(RewriterBase &rewriter, rock::GpuAllocOp allocOp,
   Location loc = allocOp->getLoc();
   OpBuilder::InsertionGuard g(rewriter);
   rewriter.setInsertionPoint(allocOp);
-  SmallVector<Value> mbAlloc(multiBufferingFactor);
+  SmallVector<rock::GpuAllocOp> mbAlloc(multiBufferingFactor);
   for (unsigned i = 0; i < multiBufferingFactor; i++) {
     mbAlloc[i] = rewriter.create<rock::GpuAllocOp>(
         loc, allocOp.getType(), ValueRange{}, allocOp->getAttrs());
   }
 
   // 3. RAUW with the particular slice, taking modular rotation into account.
-  replaceUsesAndPropagateType(rewriter, loc, allocOp, mbAlloc, candidateLoop);
+
+  SmallVector<Value> startingValues;
+  for (auto alloc : mbAlloc) {
+    startingValues.push_back(alloc);
+  }
+  replaceUsesAndPropagateType(rewriter, loc, allocOp, startingValues,
+                              candidateLoop);
 
   // 4. Finally, erase the old allocOp.
   rewriter.eraseOp(allocOp);
 
   return mbAlloc;
 }
 
-FailureOr<SmallVector<Value>>
+FailureOr<SmallVector<rock::GpuAllocOp>>
 mlir::rock::multiBuffer(rock::GpuAllocOp allocOp, unsigned multiBufferingFactor,
                         bool skipOverrideAnalysis) {
   IRRewriter rewriter(allocOp->getContext());
   return multiBuffer(rewriter, allocOp, multiBufferingFactor,
                      skipOverrideAnalysis);
 }
+
+FailureOr<SmallVector<rock::GpuAllocOp>>
+mlir::rock::updateMultiBuffer(RewriterBase &rewriter, Location loc,
+                              ArrayRef<rock::GpuAllocOp> allocs,
+                              unsigned newMultiBufferingFactor) {
+  // Select the first alloc, which will be updated
+  rock::GpuAllocOp allocOp = allocs.back();
+  OpBuilder::InsertionGuard g(rewriter);
+  rewriter.setInsertionPoint(allocOp);
+  SmallVector<GpuAllocOp> newAllocs;
+  if (allocs.size() < newMultiBufferingFactor) {
+    for (unsigned i = 0; i < newMultiBufferingFactor - allocs.size() + 1; i++)
+      newAllocs.push_back(rewriter.create<rock::GpuAllocOp>(
+          loc, allocOp.getType(), ValueRange{}, allocOp->getAttrs()));
+    SmallVector<Value> startingValues;
+    for (auto alloc : newAllocs) {
+      startingValues.push_back(alloc);
+    }
+    replaceUsesAndPropagateType(rewriter, loc, allocOp, startingValues,
+                                nullptr);
+  } else if (allocs.size() > newMultiBufferingFactor) {
+    for (unsigned i = 0; i < allocs.size() - newMultiBufferingFactor; i++)
```

Comment:
1. Why the first allocations, not the last ones?
2. Style nit ` for (unsigned i = 0, e = allocs.size() - newMultiBufferingFactor; i < e; ++i)`

Reply:
Actually I forgot to add a test for that part, which contained a bug. So, thank you! I applied your change, fixed the bug, and added some more tests